core/version/gear.ex

```diff
@@ -113,63 +115,92 @@ defmodule AntikytheraCore.Version.Gear do
     if Antikythera.Env.running_with_release?() do
       do_install_gears_at_startup(gear_names)
     else
       :ok
     end
   end
 
-  defunp do_install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok do
+  defunpt do_install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok | :error do
+    # Tests need `__MODULE__.` to mock these functions.
     gear_and_deps_pairs =
-      Enum.map(gear_names, fn g -> {g, gear_dependencies_from_app_file(g, gear_names)} end)
-
-    pairs_not_installed =
-      install_gears_whose_deps_met(gear_and_deps_pairs, MapSet.new(), fn gear_name ->
-        try do
-          install_or_upgrade_to_next_version(gear_name)
-        rescue
-          e -> L.error("Failed to install #{gear_name}: #{Exception.message(e)}")
-        end
+      Enum.map(gear_names, fn g ->
+        {g, __MODULE__.gear_dependencies_from_app_file(g, gear_names)}
       end)
 
+    {pairs_not_installed, num_failed_install} =
+      __MODULE__.install_gears_whose_deps_met(
+        gear_and_deps_pairs,
+        MapSet.new(),
+        0,
+        fn gear_name ->
+          try do
+            install_or_upgrade_to_next_version(gear_name)
+          rescue
+            e ->
+              L.error("Failed to install #{gear_name}: #{Exception.message(e)}")
+              :error
+          end
+        end
+      )
+
     Enum.each(pairs_not_installed, fn {gear_name, deps} ->
       L.error("#{gear_name} is not installed due to unmatched dependencies: #{inspect(deps)}")
     end)
+
+    num_installable = length(gear_names)
+    num_installed = num_installable - length(pairs_not_installed) - num_failed_install
+    # version_upgrade_test starts Antikythera without gears.
+    # Then, we can't check number of installed gears.
+    needs_check = Antikythera.Env.runtime_env() != :local
+
+    is_enough_gears_installed =
+      num_installable != 0 && num_installed / num_installable > @installed_gear_ratio_threshold
+
+    if !needs_check || is_enough_gears_installed do
+      :ok
+    else
+      :error
+    end
   end
 
-  defunpt install_gears_whose_deps_met(
-            pairs :: v[[gear_and_deps_pair]],
-            installed_gears_set :: MapSet.t(),
-            f :: (GearName.t() -> :ok)
-          ) :: [gear_and_deps_pair] do
+  # public for mock
+  defun install_gears_whose_deps_met(
+          pairs :: v[[gear_and_deps_pair]],
+          installed_gears_set :: MapSet.t(),
+          num_failed_install :: v[non_neg_integer],
+          f :: (GearName.t() -> :ok | :error)
+        ) :: {[gear_and_deps_pair], non_neg_integer} do
     if Enum.empty?(pairs) do
-      []
+      {[], num_failed_install}
     else
       {pairs_installable, pairs_not_installable} =
         Enum.split_with(pairs, fn {_, deps} -> MapSet.subset?(deps, installed_gears_set) end)
 
       if Enum.empty?(pairs_installable) do
         # we cannot make progress any more
-        pairs_not_installable
+        {pairs_not_installable, num_failed_install}
```

Comment:
**(optional)**

If the following error log is written here, the return value of this function will be simple (the number of gears that could not be installed).

```elixir
    Enum.each(pairs_not_installed, fn {gear_name, deps} ->
      L.error("#{gear_name} is not installed due to unmatched dependencies: #{inspect(deps)}")
    end)
```

Reply:
```
the current code may be more appropriate because both error loggings are in do_install_gears_at_startup/1.
```

I think this is appropriate.
`install_gears_whose_deps_met/4` is something like iterator, and `do_install_gears_at_startup/1` does the actual work i.e. installing gears and reporting errors.